### PR TITLE
feat(rocjitsu): post-routing memory access hook

### PR DESCRIPTION
**Pull request by amd-arosa**
## Motivation

`onAmdgpuRouteMemoryInstruction` fires at the top of
`ComputeUnitCore::route_memory_inst`, before routing has decided anything. For
one important case that is not what the memory system is asked for.

A FLAT access whose addresses land in the shared aperture decodes as global and
is issued to the *local* pipeline, with every lane's address rewritten into the
workgroup's LDS allocation and its wait counter changed from VMCNT to LGKMCNT.
An observer looking before routing charges that access to the vector cache, at
an address the memory system never uses, and waits on a counter no `s_waitcnt`
in the kernel names.

## Technical Details

`ExecutionPlugin::onAmdgpuMemoryAccessRouted(const MemoryAccessObservation &)`
fires once routing has settled: after the aperture rewrite, before the pipeline
takes the instruction. That is the one point at which the space, the wait
counter and the addresses are all the ones about to be used.

`MemoryAccessObservation` carries facts, not estimates:

- Complete identity -- mnemonic, PC, compute unit, dispatch, queue, workgroup,
  wavefront slot, VMID -- so a consumer can compare two observations, or an
  observation against another event, field for field.
- The route it was issued to, and separately whether it was *rewritten* into
  LDS. Separating "traffic that is really LDS" from "traffic that became LDS"
  needs both.
- Four lane masks. Active is what the instruction issued with, usually EXEC but
  not always. Valid is the lanes whose address resolved. Request is the lanes
  that produce a transaction, which differs for the wave64 transpose loads that
  issue through the low half. Scratch is the lanes whose addresses are
  swizzled, which a FLAT wave can mix with global ones. Inactive and unknown
  are derived rather than stored, so they cannot drift.
- Per-lane addresses, per-element lane validity, and the second address set of
  a DS dual access.

Three facts are taken from the wavefront rather than the instruction state,
because the state is the pipeline's and the pipeline has not run: the lane
count, the transpose request mask that keys on it, and the queue id.
`VectorMemState::wf_size` defaults to 64 whatever the target and is set on some
paths only after routing, so a wave32 access reaching the hook can still be
carrying 64. `transpose_request_lane_mask` gains an overload taking the width
explicitly; existing callers are unchanged.

An instruction no pipeline accepts is reported with an `UNKNOWN` route rather
than dropped, so a consumer counting a kernel's memory traffic can see there
was something it cannot account for.

The observation is built only when a plugin opts into the hook, through the
same per-hook mechanism the group already samples for scalar-register reads:
building it is real work on the per-instruction path, and every plugin in the
tree but one has no use for it.

`AtomicOp` moves to a header of its own so the observation -- and so
`execution_plugin.h`, and so every plugin translation unit -- no longer pulls
in `mem_state.h` and the headers behind it.

Note the new virtual sits mid-class, shifting every later vtable slot. The
plugin loader is deliberately unversioned, so a partially rebuilt out-of-tree
plugin calls the wrong hook with no diagnostic; plugins must be rebuilt with
the library.

## Issue Tracking

#11119

## Test Plan

added 12 test cases covering the spaces and edge cases the plan names. The
central one asserts a FLAT access to the shared aperture against what the
*pre*-routing hook was shown in the same run -- global, VMCNT, aperture address
-- versus what the memory system is asked for -- local, LGKMCNT, LDS address.
Every field the observation carries was mutation-checked.

## Test Result

rocjitsu_tests: 3836 passed, 2 skipped, 0 failed.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/atomic_op.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+/// @file atomic_op.h
+/// @brief The atomic read-modify-write operations a memory instruction can name.
+
+#ifndef ROCJITSU_VM_AMDGPU_ATOMIC_OP_H_
+#define ROCJITSU_VM_AMDGPU_ATOMIC_OP_H_
+
+#include <cstdint>
+
+namespace rocjitsu {
+namespace amdgpu {
```
Comment (atgutier) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/atomic_op.h]: This doesn't need to be amdgpu. RISC-V also has these same atomics.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/atomic_op.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  UMAX,           ///< Unsigned maximum.
+  AND,            ///< Bitwise AND.
+  OR,             ///< Bitwise OR.
+  XOR,            ///< Bitwise XOR.
+  INC,            ///< Increment (wrapping).
+  DEC,            ///< Decrement (wrapping).
+  FADD,           ///< Floating-point add.
+  FMIN,           ///< Floating-point minimum.
+  FMAX,           ///< Floating-point maximum.
+  APPEND,         ///< LDS append counter.
```
Comment (atgutier) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/atomic_op.h]: While you're moving this can you add a todo to move the LDS pieces out and into and LDS-specific header? Or take the opportunity to do that yourself here :-).

**Review by atgutier — APPROVED** [review]
